### PR TITLE
Add improved CI

### PR DESCRIPTION
**Pull request by szepeviktor**
# Description

- Checks for many-many problems (all resolved in this PR)
- Removes **huge** amount of mixed code styling ❌ See [changes without whitespaces](https://github.com/infinum/eightshift-libs/pull/332/files?diff=unified&w=1)
- The distributed package got much smaller
- Fixes many problems, see commit messages please
- Was fun 🎈 to make

# Screenshots / Videos

kép

# Linked documentation PR

All my knowledge 🧠 

_(this is my standard CI config)_

### CODE REVIEW

**Comment by szepeviktor** [comment]
BTW There are a lot of PSR-4 loading problems in tests.

May I send another PR?

```
Class Tests\Unit\Columns\TaxonomyColumnMock located in ./tests/Unit/Columns/TaxonomyColumnsTest.php does not comply with psr-4 autoloading standard. Skipping.
Class Tests\Unit\Columns\MediaColumnMock located in ./tests/Unit/Columns/Media/MediaColumnsTest.php does not comply with psr-4 autoloading standard. Skipping.
Class Tests\Unit\Columns\UserColumnMock located in ./tests/Unit/Columns/UsersColumnsTest.php does not comply with psr-4 autoloading standard. Skipping.
Class Tests\Unit\Columns\PostTypeColumnMock located in ./tests/Unit/Columns/PostTypeColumnsTest.php does not comply with psr-4 autoloading standard. Skipping.
Class Tests\Unit\Cli\AbstractTest located in ./tests/Unit/Cli/AbstractCliTest.php does not comply with psr-4 autoloading standard. Skipping.
Class Tests\Unit\Helpers\MockErrorLogger located in ./tests/Unit/Helpers/ErrorLoggerTraitTest.php does not comply with psr-4 autoloading standard. Skipping.
Class Tests\Unit\Helpers\MockObjectHelper located in ./tests/Unit/Helpers/ObjectHelperTraitTest.php does not comply with psr-4 autoloading standard. Skipping.
Class Tests\Unit\Main\MainTest located in ./tests/Unit/Main/AbstractMainTest.php does not comply with psr-4 autoloading standard. Skipping.
Class Tests\Unit\Main\MainCompiledTest located in ./tests/Unit/Main/AbstractMainTest.php does not comply with psr-4 autoloading standard. Skipping.
Class Tests\Datasets\Autowiring\Invalid\DoesNotExist\ServiceWithInvalidNamespace located in ./tests/Datasets/Autowiring/Invalid/ServiceWithInvalidNamespace.php does not comply with psr-4 autoloading standard. Skipping.
```

**Review by dingo-d — CHANGES_REQUESTED** [review]
Love the additions! Thank you for the work on improving the codebase.

I have left some questions and suggestions in the comments.

Just to ask here: Is it possible to have two separate workflows but have one depend on another like in the `needs` part? I love to separate concerns, make the workflow file a bit easier to read than one massive file (although the current one you propose does have huge improvement).

File: .github/workflows/integrate.yml [hunk 1]
```diff
@@ -0,0 +1,178 @@
+# yaml-language-server: $schema=https://json.schemastore.org/github-workflow
+
+name: "Integrate"
```
[the remark this answers is not in the document]
Comment (szepeviktor) [reply .github/workflows/integrate.yml:3]: "CI" ==== continuous >integrat<ion

File: .github/workflows/integrate.yml [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+      - name: "Checkout code"
+        uses: "actions/checkout@v3"
+
+      - name: "Check file permissions"
+        run: |
+          test "$(find . -type f -not -path './.git/*' -executable)" = ""
+
+      - name: "Find non-printable ASCII characters"
+        run: |
+          ! LC_ALL=C.UTF-8 find . -type f -name "*.php" -print0 | xargs -0 -- grep -PHn "[^\t -~]"
```
[the remark this answers is not in the document]
Comment (szepeviktor) [reply .github/workflows/integrate.yml:34]: This check for strings like `"Žoljom"` which is valid in all programing langues.

File: .github/workflows/integrate.yml [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        uses: "ramsey/composer-install@v2"
+        with:
+          dependency-versions: "highest"
+
+      - name: "Execute static analysis"
+        run: "composer run test:types"
+        # @TODO Magic Number Detector, Copy-Paste Detector
+
+  coding_standards:
+    name: "4️⃣ Coding Standards"
```
[the remark this answers is not in the document]
Comment (szepeviktor) [reply .github/workflows/integrate.yml:132]: @dingo-d So you would like to add the word "checks" seven times to this UI?

kép

[the remark this answers is not in the document]
Comment (szepeviktor) [reply .github/workflows/integrate.yml:132]: Jobs are containers for steps, they don't do anything.
Steps in a job do different things, one of them is a check.

**Comment by szepeviktor** [comment]
> Is it possible to have two separate workflows but have one depend on another like in the needs part?

```yaml
on:
  workflow_run:
    workflows:
      - "Build"
    types:
      - "completed"

jobs:
  step-name:
    if: "${{ github.event.workflow_run.conclusion == 'success' }}"
    steps:
      - run: |
          echo 'The triggering workflow passed'
```

📖 https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#running-a-workflow-based-on-the-conclusion-of-another-workflow

File: .github/workflows/integrate.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (dingo-d) [reply .github/workflows/integrate.yml:34]: Yes, but the check is for non printable ASCII characters, which sounds like a syntax check to me 🤷🏼‍♂️

File: .github/workflows/integrate.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (dingo-d) [reply .github/workflows/integrate.yml:132]: Yeah, the above would read nicer with checks to me (except for the tests one ofc).

**Comment by dingo-d** [comment]
> > Is it possible to have two separate workflows but have one depend on another like in the needs part?
> 
> Yes, it is. It makes very hard to understand CI.
> 
> ```yaml
> on:
>   workflow_run:
>     workflows:
>       - "Build"
>     types:
>       - "completed"
> 
> jobs:
>   my-job-id:
>     if: "${{ github.event.workflow_run.conclusion == 'success' }}"
>     steps:
>       - run: |
>           echo 'The triggering workflow passed'
> ```
> 
> 📖 https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#running-a-workflow-based-on-the-conclusion-of-another-workflow

Yeah, I think that would make things a bit harder to understand. We can leave the one main file 👍🏼

File: .editorconfig [hunk 4]
```diff
@@ -0,0 +1,31 @@
+# https://EditorConfig.org
```
[the remark this answers is not in the document]
Comment (mbmjertan) [reply .editorconfig:1]: A basic `.editorconfig` would be great to have IMO, as we could set default settings for indentation. Other coding style settings can be set in the IDE directly, but this is something that's already mandated on the project level with coding standards, so I see no resason why not to provide an `.editorconfig`

File: .github/workflows/integrate.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (mbmjertan) [reply .github/workflows/integrate.yml]: It is a syntax check, but not a syntax error 🤷

**Comment by szepeviktor** [comment]
All your wishes came true!

File: .editorconfig [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+[*]
+indent_style = tab
+indent_size = 4
+end_of_line = lf
+charset = utf-8
+trim_trailing_whitespace = true
+insert_final_newline = true
+block_comment_start = /*
+block_comment = *
+block_comment_end = */
```
[the remark this answers is not in the document]
Comment (szepeviktor) [reply .editorconfig:14]: You're welcome.

File: .github/workflows/integrate.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (szepeviktor) [reply .github/workflows/integrate.yml]: This is a low-level stuff, "Byte-level" how I call it.
I've moved this step to "Syntax errors checks"

**Review by dingo-d — APPROVED** [review]
Amazing, thank you very much for these improvements 👍🏼

**Review by mbmjertan — APPROVED** [review]
Amazing! Thank you for your contribution 🎉 😄 🙇

**Pull request merged**